# **HOMEWORK 3...?** 🕵
## **DUE: *MARCH 25, 2025 @ 11:59 PM***
## **24-HR LATE DUE DATE W/ A 15% PENALTY: *MARCH 26, 2025 @ 11:59 PM***

### **DO NOT REMOVE ANY PART OF ANY OF THE QUESTIONS OR YOU LOSE CREDIT**
### *No Hardcoding Any Part of Any Question*  😋🪄🧌
### **REMEMBER TO SHOW ALL CODE OUTPUT (NO CREDIT OTHERWISE)**
### **Reminder: Please make sure your code runs before submitting your work. Code sections that do not run will receive 0 credits, no partials will be given. This is VERY important in real project development.**

---
---
**DISCLAIMER: This is a fake story, not a real incident! We have created this story just for the data exploration purpose.**

---
---

**Campus Police Department**

Incident Report

---

**Incident Date:** Febuary 27, 2025

**Location of Incident:** 2nd Floor, Iribe Building

**Items Left at the Scene:** Red Ticket for Zichao's Magic Show

**Case Number:** 2025-IRB-0027

---
**Victim Information:**

**Name:** Dr. Fardina Alam

**Affiliation:** University of Maryland, College Park

**Position:** Professor

**Contact Information:** fardina@umd.edu

---

**Incident Description:**

On the evening of Febuary 27, 2025, an unknown individual(s) unlawfully entered Dr. Fardina's room located on the 2nd floor of Iribe and stole a set of Markers. The stolen items were a set of limited edition markers that Dr. Fardina had spent years collecting. At the scene of the crime, a red entry ticket for "Zichao's Magic Show" was found on the floor near Dr. Fardina's desk.






In [3]:
# We import all essential libraries for this HW.
# If you want to add additional libraries, you can add more here!
import pandas as pd
from datetime import datetime
import string

## YOUR TASK 🔎:

In this assignment, you are tasked with solving the mystery of Dr. Fardina's stolen markers and finding the culprit.

* To accomplish this, you will be provided with **4 data sets** containing key information related to the incident. The list of the data sets name given below:

 - Profiles
 - Witnesses
 - Zichao's magic show
 - Interrogation Statements



* Your goal is to analyze these data sets to uncover the clues and evidences that will lead you to the thief.

**Be careful in following instructions as it may lead to you losing points :(**

Take a moment to look at all the data sets below:

---

**Profiles:**

This dataset contains information regarding all the potential people that could have committed the crime. General information regarding each person are as follows:
- **ID:** A **unique** identification for each person
- **Name:** The name of the person
- **Birthday:** The birthday of the person
- **Location_during_crime:** The location of the person during the crime
- **Eye_Color:** The eye color of the person
- **Skin_Color:** The skin color of the person
- **Hair_Color:** The hair color of the person
- **Height:** The height of the person in cm

**Witnesses:**

This dataset contains a statement from everyone in profiles, regarding suspicious things they saw during the time of the crime. Information that comes in this dataset are as follows:
- **ID:** A **Unique** identification for each person
- **Statement:** A statement from the person regarding a suspicious activity during the time of the crime

**Zichao's Magic Show:**

This dataset contains everyone that purchased a ticket for the show. **NOTE** that not everyone in profiles has purchased a ticket for this show. The Information contained in this dataset are as follows:
- **ID:** A **Unique** indentification for each person
- **Ticket_Number:** A **Unique** number that serves as an identification for the ticket bought

**Interrogation Statements:**

This datasets contains a statement from each person after they have been interrogated. **NOTE** assume that each statement is ***100% true*** and that ***there are people in this dataset who aren't in the "Profiles" dataset***.
Information contained in this dataset are as follows:

- **Name:** The name of the person being interrogated
- **Statement:** A statement the person gives after being interrogated

---

## Q1) Lets begin by observing the profiles dataset (10 PTS Total)






**TASK 1.1 (1 Point):**  In the code box below, load in "Profiles" dataset into a variable called **profiles_df** and display it

In [4]:
profiles_df = pd.read_csv('Profiles.csv')
profiles_df

ID              Name    Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi  1969/07/10          fourth floor       red   
1    P086       Aarav Patel  1979/03/10          second floor    indigo   
2    P022            Li Wei  1990/10/08          fourth floor      blue   
3    P003     Haruto Tanaka  1996/27/10                 lobby     green   
4    P036      Kwame Mensah  1991/26/02          fourth floor      blue   
..    ...               ...         ...                   ...       ...   
343  P126       Andrew Choe  1988/03/01                   NaN    violet   
344  P203     Krish Thakker  1965/26/09           third floor    yellow   
345  P273     Jeonghan Yoon  2003/12/07          fourth floor      blue   
346  P290  Ananya Malipeddi  10/03/1994           third floor       NaN   
347  P233         Ifra Syed  06/27/1982          second floor       red   

    Skin_Color Hair_Color  Height  
0       indigo     orange     191  
1       indigo      green     198  
2          red     violet     154  
3         blue     violet     158  
4       yellow     orange     184  
..         ...        ...     ...  
343     indigo        red     176  
344     indigo     orange     193  
345     indigo        red     193  
346     yellow     indigo     161  
347     yellow     orange     169  

[348 rows x 8 columns]

**TASK 1.2 (1 Point):**  Next, you need to display the number of rows and columns in the "Profiles" dataset.

In [5]:
profiles_df.shape

(348, 8)

**TASK 1.3 (1 Point):**  Next, you need to display the name of the columns in the "Profiles" dataset.

In [6]:
profiles_df.columns

Index(['ID', 'Name', 'Birthday', 'Location_during_crime', 'Eye_Color',
       'Skin_Color', 'Hair_Color', 'Height'],
      dtype='object')

**TASK 1.4 (1 Point):**  Now, you need to display the types of the columns in the "Profiles" dataset.

In [7]:
profiles_df.dtypes

ID                       object
Name                     object
Birthday                 object
Location_during_crime    object
Eye_Color                object
Skin_Color               object
Hair_Color               object
Height                    int64
dtype: object

**TASK 1.5 (2 Points):** Utilizing the `count()` function on this dataset and set it to the variable **info** and then display it.

Write any **potential problem** that you see with the dataset when observing the output of the count function:




In [8]:
info = profiles_df.count()
info

ID                       348
Name                     348
Birthday                 348
Location_during_crime    280
Eye_Color                280
Skin_Color               348
Hair_Color               348
Height                   348
dtype: int64

**<font color='red'>Write your observations here:</font>**

According to the count function, there are 348 recorded profiles in the dataset, but only 280 of them have a recorded location during the time of the crime and also only 280 of them have a recorded eye color, so there is missing data.

**TASK 1.6 (4 Points):**

Look back at the "Profiles" dataset. Using other pandas functions than `count()`, **can you spot another potential problem with the dataset?**

In [9]:
print("Full Dataframe:")
print(profiles_df)

# Check for missing values in each column
missing_values = profiles_df.isnull().sum()
print("Missing values in each column:")
print(missing_values)

# Check for duplicate rows
duplicate_rows = profiles_df.duplicated().sum()
print(f'\nNumber of duplicate rows: {duplicate_rows}')

# Get summary statistics for each column
summary_statistics = profiles_df.describe()
print("\nSummary statistics for each column:")
print(summary_statistics)

Full Dataframe:
       ID              Name    Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi  1969/07/10          fourth floor       red   
1    P086       Aarav Patel  1979/03/10          second floor    indigo   
2    P022            Li Wei  1990/10/08          fourth floor      blue   
3    P003     Haruto Tanaka  1996/27/10                 lobby     green   
4    P036      Kwame Mensah  1991/26/02          fourth floor      blue   
..    ...               ...         ...                   ...       ...   
343  P126       Andrew Choe  1988/03/01                   NaN    violet   
344  P203     Krish Thakker  1965/26/09           third floor    yellow   
345  P273     Jeonghan Yoon  2003/12/07          fourth floor      blue   
346  P290  Ananya Malipeddi  10/03/1994           third floor       NaN   
347  P233         Ifra Syed  06/27/1982          second floor       red   

    Skin_Color Hair_Color  Height  
0       indigo     orange     191  
1       ind

**<font color='red'>Write your observations here:</font>**

After observing the entire dataframe, apart from the missing values, I can see that the last two profiles have birthday dates that are in an inconsistent format with the rest.

## Q2) Cleaning the Dataset (25 PTS Total)

**TASK 2.1 (10 Points):** Lets begin by fixing the Birthday column. You may notice that the dates are in inconsistent format!

- Pandas internally represent dates in 'datetime64' format which is a form of ISO 8601. Basically, dates are represented as yyyy-mm-dd.


- Fix all the dates in the Birthday Column.
- Display the "Profiles" dataframe after you are done.
- Note that the only **2 formats** the dates are in is either **yyyy/dd/mm** or **mm/dd/yyyy**;
- You want to represent them as **yyyy-mm-dd**

- **EXAMPLE**: In the “Profiles” dataset, dates like *2000/10/12* and *10/12/2000* follow different formats—one as *December 10th, 2000*, and the other as *October 12th, 2000*. We aim to standardize them to *2000-12-10* and *2000-10-12*, respectively.

**HINT:** some datetime functions might make things easier.
- https://www.geeksforgeeks.org/python-pandas-to_datetime/
- https://www.programiz.com/python-programming/datetime/strftime


In [10]:
profiles_df['Birthday'] = pd.to_datetime(profiles_df['Birthday'], errors='coerce', format='%Y/%d/%m').fillna(pd.to_datetime(profiles_df['Birthday'], errors='coerce', format='%m/%d/%Y'))

profiles_df['Birthday'] = profiles_df['Birthday'].dt.strftime('%Y-%m-%d')

profiles_df

ID              Name    Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi  1969-10-07          fourth floor       red   
1    P086       Aarav Patel  1979-10-03          second floor    indigo   
2    P022            Li Wei  1990-08-10          fourth floor      blue   
3    P003     Haruto Tanaka  1996-10-27                 lobby     green   
4    P036      Kwame Mensah  1991-02-26          fourth floor      blue   
..    ...               ...         ...                   ...       ...   
343  P126       Andrew Choe  1988-01-03                   NaN    violet   
344  P203     Krish Thakker  1965-09-26           third floor    yellow   
345  P273     Jeonghan Yoon  2003-07-12          fourth floor      blue   
346  P290  Ananya Malipeddi  1994-10-03           third floor       NaN   
347  P233         Ifra Syed  1982-06-27          second floor       red   

    Skin_Color Hair_Color  Height  
0       indigo     orange     191  
1       indigo      green     198  
2          red     violet     154  
3         blue     violet     158  
4       yellow     orange     184  
..         ...        ...     ...  
343     indigo        red     176  
344     indigo     orange     193  
345     indigo        red     193  
346     yellow     indigo     161  
347     yellow     orange     169  

[348 rows x 8 columns]

**TASK 2.2 (5 Points):** Now that we have cleaned up the Birthday column, lets find the age of each person. Create a new column called "Age" that displays the age of the person **on the day of the crime.** Once you are done, display the
"Profiles" dataframe

In [11]:
profiles_df['Birthday'] = pd.to_datetime(profiles_df['Birthday'], errors='coerce')

date_of_crime_str = "2025-02-27"
date_of_crime = pd.to_datetime(date_of_crime_str)

profiles_df['Age'] = date_of_crime.year - profiles_df['Birthday'].dt.year - ((date_of_crime.month < profiles_df['Birthday'].dt.month) | ((date_of_crime.month == profiles_df['Birthday'].dt.month) & (date_of_crime.day < profiles_df['Birthday'].dt.day)))

profiles_df

ID              Name   Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi 1969-10-07          fourth floor       red   
1    P086       Aarav Patel 1979-10-03          second floor    indigo   
2    P022            Li Wei 1990-08-10          fourth floor      blue   
3    P003     Haruto Tanaka 1996-10-27                 lobby     green   
4    P036      Kwame Mensah 1991-02-26          fourth floor      blue   
..    ...               ...        ...                   ...       ...   
343  P126       Andrew Choe 1988-01-03                   NaN    violet   
344  P203     Krish Thakker 1965-09-26           third floor    yellow   
345  P273     Jeonghan Yoon 2003-07-12          fourth floor      blue   
346  P290  Ananya Malipeddi 1994-10-03           third floor       NaN   
347  P233         Ifra Syed 1982-06-27          second floor       red   

    Skin_Color Hair_Color  Height  Age  
0       indigo     orange     191   55  
1       indigo      green     198   45  
2          red     violet     154   34  
3         blue     violet     158   28  
4       yellow     orange     184   34  
..         ...        ...     ...  ...  
343     indigo        red     176   37  
344     indigo     orange     193   59  
345     indigo        red     193   21  
346     yellow     indigo     161   30  
347     yellow     orange     169   42  

[348 rows x 9 columns]

**TASK 2.3 (10 Points):** Now that we have created the "Age"; lets clean the 'Eye_color's column.

***Please read the directions carefully:***


- In order to fill in the missing eye_colors, we want to find the **most common eye_color** among those who have the **same hair and skin color**.
- For Instance, if we want to find the eye_color of someone who has yellow hair and violet skin:
 - we would first find everything who has the same hair and skin color, in this case it would be yellow hair and violet skin
 - and then we find the most common eye_color among them.
 - we then replace the missing eye_color with the eye color that is the most common among the people with the same hair and skin color.
- Watch out for redundant data!

- ***NOTE: In the case that you find multiple modes, choose the color that appears first in alphabetical order***

 When you finish filling in the missing eye colors, please **display the profiles_df**


In [38]:
from collections import Counter

def most_common_eye_color(group):
    eye_colors = group.dropna().tolist()

    if not eye_colors:
        return None
    
    eye_color_count = Counter(eye_colors)
    most_common_eye_color = eye_color_count.most_common()
    max_count = most_common_eye_color[0][1]
    common_colors = [color for color, count in most_common_eye_color if count == max_count]

    return sorted(common_colors)[0]

common_eye_colors = profiles_df.groupby(['Hair_Color', 'Skin_Color'])['Eye_Color'].apply(most_common_eye_color).reset_index()
common_eye_colors = common_eye_colors.rename(columns={'Eye_Color': 'Most_Common_Eye_Color'})
profiles_df = profiles_df.merge(common_eye_colors, on=['Hair_Color', 'Skin_Color'])
profiles_df['Eye_Color'] = profiles_df['Eye_Color'].fillna(profiles_df['Most_Common_Eye_Color'])
profiles_df = profiles_df.drop(columns=['Most_Common_Eye_Color'])

profiles_df

ID              Name   Birthday Location_during_crime Eye_Color  \
0    P299       Marco Rossi 1969-10-07          fourth floor       red   
1    P086       Aarav Patel 1979-10-03          second floor    indigo   
2    P022            Li Wei 1990-08-10          fourth floor      blue   
3    P003     Haruto Tanaka 1996-10-27                 lobby     green   
4    P036      Kwame Mensah 1991-02-26          fourth floor      blue   
..    ...               ...        ...                   ...       ...   
343  P126       Andrew Choe 1988-01-03                   NaN    violet   
344  P203     Krish Thakker 1965-09-26           third floor    yellow   
345  P273     Jeonghan Yoon 2003-07-12          fourth floor      blue   
346  P290  Ananya Malipeddi 1994-10-03           third floor    yellow   
347  P233         Ifra Syed 1982-06-27          second floor       red   

    Skin_Color Hair_Color  Height  Age  
0       indigo     orange     191   55  
1       indigo      green     198   45  
2          red     violet     154   34  
3         blue     violet     158   28  
4       yellow     orange     184   34  
..         ...        ...     ...  ...  
343     indigo        red     176   37  
344     indigo     orange     193   59  
345     indigo        red     193   21  
346     yellow     indigo     161   30  
347     yellow     orange     169   42  

[348 rows x 9 columns]


## Q3) Finding the features of the culprit (38 PTS Total)

Now that we have successfully cleaned all of our data, lets begin finding the features of the criminal. Take a moment to look at the **"Witnesses"** table. We don't necessarily want all the statements from every single witness.

**NOTE:** Take a look at the structure of each statement, notice how there are only a few variations of statements from the witnesses.

**TASK 3.1 (5 Points):** Find all witnesses that were near the crime scene
- First load in the Witness.csv file and set it to a variable called "**witness_df**"
- create a new dataframe called **"important_witnesses"** that contains the id and the witness statements of those who were near the scene of the crime
- display the new dataframe when you are finished
- **HINT:** look at the floor which the crime happened on.

In [63]:
witness_df = pd.read_csv('Witness.csv')

crime_location = "second floor"
witness_profiles_df = witness_df.merge(profiles_df, on='ID', how='left')
important_witnesses = witness_profiles_df[witness_profiles_df['Location_during_crime'] == crime_location][['ID', 'Statement']]

important_witnesses

ID                                          Statement
13   P150  I saw someone with orange hair acting suspicious.
17   P075  I saw someone with orange hair acting suspicious.
21   P058   I saw someone with green eyes acting suspicious.
22   P157  I saw someone who looked 152 cm tall acting su...
23   P157  I saw someone who looked 152 cm tall acting su...
..    ...                                                ...
329  P077  I saw someone with orange eyes acting suspicious.
330  P168     I saw someone with red skin acting suspicious.
331  P168     I saw someone with red skin acting suspicious.
345  P145  I saw someone with orange eyes acting suspicious.
346  P145  I saw someone with orange eyes acting suspicious.

[71 rows x 2 columns]

**TASK 3.2 (10 Points):** Find every potential eye color that the culprit could have

 - Put all the potential eye colors of the culprit in a list called **"potential_eye_colors"** and **display** it when you're done.
 - **potential_eye_colors** should have **no duplicates.**

In [237]:
valid_eye_colors = profiles_df['Eye_Color'].dropna().unique()

potential_eye_colors = []

for statement in important_witnesses['Statement']:
    statement_in_lowercase = statement.lower()
    
    for eye_color in valid_eye_colors:
        expected_statement_with_eye_color = f'i saw someone with {eye_color.lower()} eyes acting suspicious.'

        if  expected_statement_with_eye_color == statement_in_lowercase:
            potential_eye_colors.append(eye_color)

potential_eye_colors = list(set(potential_eye_colors))

print(f"Potential Eye Colors: {potential_eye_colors}")

Potential Eye Colors: ['green', 'blue', 'orange', 'violet']


**TASK 3.3 (5 Points):** Find every potential hair color that the culprit could have

 - Put all the potential hair colors of the culprit in a list called **"potential_hair_colors"** and **display** it when you're done.

In [258]:
valid_hair_colors = profiles_df['Hair_Color'].dropna().unique()

potential_hair_colors = []

for statement in important_witnesses['Statement']:
    statement_in_lowercase = statement.lower()
    
    for hair_color in valid_hair_colors:
        expected_statement_with_hair_color = f'i saw someone with {hair_color.lower()} hair acting suspicious.'

        if  expected_statement_with_hair_color == statement_in_lowercase:
            potential_hair_colors.append(hair_color)

potential_hair_colors = list(set(potential_hair_colors))

print(f"Potential Hair Colors: {potential_hair_colors}")

Potential Hair Colors: ['green', 'blue', 'orange', 'violet']


**TASK 3.4 (5 Points):** Find every potential skin color that the culprit could have

 - Put all the potential skin colors of the culprit in a list called **"potential_skin_colors"** and display it when you're done

In [268]:
valid_skin_colors = profiles_df['Skin_Color'].dropna().unique()

potential_skin_colors = []

for statement in important_witnesses['Statement']:
    statement_in_lowercase = statement.lower()
    
    for skin_color in valid_skin_colors:
        expected_statement_with_skin_color = f'i saw someone with {skin_color.lower()} skin acting suspicious.'

        if  expected_statement_with_skin_color == statement_in_lowercase:
            potential_skin_colors.append(skin_color)

potential_skin_colors = list(set(potential_skin_colors))

print(f"Potential Skin Colors: {potential_skin_colors}")

Potential Skin Colors: ['green', 'orange', 'red']


**TASK 3.5 (10 Points):** Find the potential height of the culprit

 - Put the smallest and largest potential heights of the culprit in the variables called **"smallest_height" & "largest_height"** and display them when you are done

In [307]:
potential_heights = []

for statement in important_witnesses['Statement']:
    statement_in_lowercase = statement.lower()
    
    if "cm tall" in statement_in_lowercase:
        height_part = statement_in_lowercase.split("cm tall")[0]
        height = height_part.split("looked")[-1].strip()
        
        if height.isdigit():
            potential_heights.append(int(height))

smallest_height = min(potential_heights)
largest_height = max(potential_heights)

print(f"Smallest Height: {smallest_height}")
print(f"Largest Height: {largest_height}")

Smallest Height: 150
Largest Height: 170


**TASK 3.6 (8 Points):** Find the potential age of the culprit

 - Put the smallest and largest potential age of the culprit in the variables called **"smallest_age" & "largest_age"** and display them when you are done

In [308]:
potential_ages = []

for statement in important_witnesses['Statement']:
    statement_in_lowercase = statement.lower()
    
    if "years old" in statement_in_lowercase:
        age_part = statement_in_lowercase.split("years old")[0]
        age = age_part.split("looks to be")[-1].strip()
        
        if age.isdigit():
            potential_ages.append(int(age))

smallest_age = min(potential_ages)
largest_age = max(potential_ages)

print(f"Smallest Age: {smallest_age}")
print(f"Largest Age: {largest_age}")

Smallest Age: 21
Largest Age: 35


## Q4) Who owns the Ticket? (30 PTS Total)

**TASK 4.1 (2 Points):** Loading final dfs


Load in the **"Zichao's Magic Show.csv"** into a dataframe called **"show_df"** and the **"Interrogation_Statement.csv"** into a dataframe called **"interrogation_df"** respectively and display them when you are done

In [329]:
show_df = pd.read_csv("Zichao_s Magic Show.csv")
interrogation_df = pd.read_csv("Interrogation_Statements.csv")

print("Zichao's Magic Show DataFrame:")
display(show_df)

print("Interrogation Statements DataFrame:")
display(interrogation_df)

Zichao's Magic Show DataFrame:


ID  Ticket_Number
0    P132       50269975
1    P056       28120732
2    P250       88051289
3    P197       79545169
4    P012       83521845
..    ...            ...
146  P121       31341288
147  P003       85008158
148  P180       24491860
149  P283       38524355
150  P142        2238676

[151 rows x 2 columns]

Interrogation Statements DataFrame:


Name                                          Statement
0    Kristopher Stoichkov  "Re%spe%cti%ng o%the%rs' be%lo%ngi%ngs i%s my ...
1           Naman Nagelia  "I@'d ne@ve@r je@o@pa@rdi@ze@ my re@pu@ta@ti@o...
2    Pranavh Vallabhaneni  "My pro^je^ct pre^se^nta^ti^o^n o^ccu^pi^e^d m...
3               Stacy Sun  "Thre$e$ co$nse$cu$ti$ve$ e$xa$ms dra$i$ne$d m...
4          Rina Kobayashi  "I% wa%s o%ff-ca%mpu%s a%tte%ndi%ng a% fa%mi%l...
..                    ...                                                ...
495             Ishan Kar  "I$ sto$le$ i$t i$n 2024 a$nd bri$be$d the$ po...
496          Adrian Black  "I% wa%s bu%sy wi%th my pro%je%ct pre%se%nta%t...
497       Brittany Taylor  "I@ wa@s bu@sy wi@th my gro@u@p pro@je@ct me@e...
498        William Miller  "I^ wa^sn't e^ve^n o^n ca^mpu^s tha^t da^y. Ch...
499  Christopher Ferguson  "Ste$a$li$ng? Tha$t's no$t i$n my na$tu$re$ a$...

[500 rows x 2 columns]

We're **so close** to figuring out the *last* and *final* clue to locate the culprit, but it looks like they are trying to beat us to it! What a devious criminal!

The culprit must have **sabotaged** our data after **interrogation** because it looks so wonky with weird symbols.

However, HQ informed us that our good friend and exemplar cryptanalyst **Alan Turing** has something to say to help us fix the data. **He encoded his message to keep it safe from the culprit.**

**TASK 4.2 (10 Points):**

Find what **Alan Turing** has to say:
- Use the **interrogation_df** to locate Alan's Turing's message.
- Put the message in a variable called "**encoded_message**" as a string (**Exclude** any **quotation marks** within the statement)
 - **Print** the encoded message in the following format: {"Encoded Message:", encoded_message}
- Figure out what the message says by taking a look at the numbers:
 - A **single space** between numbers indicates a letter to form a word.
 - A **double space** between numbers indicates the start of a new word.
- Store his decoded message in a variable called "**decoded_message**"
- Print out **Alan Turing's** decoded message in the following format: {"Decoded Message:", decoded_message}

**Hint:** Maybe [enumerating](https://docs.python.org/3/library/functions.html#enumerate) to store letters to numbers might help along with using [ascii values](https://docs.python.org/3/library/string.html)

In [553]:
encoded_message = interrogation_df.loc[interrogation_df['Name'] == 'Alan Turing', 'Statement'].values[0]
encoded_message = encoded_message.strip('"')
print({"Encoded Message:", encoded_message})

ascii_mapping = {index: chr(index) for index, _ in enumerate(range(32, 127), start=32)}

words = encoded_message.split("  ")

decoded_message = ""

for word in words:
    numbers = word.split(" ")
    decoded_word = ""
    
    for num in numbers:
        decoded_letter = ascii_mapping[int(num) + 96]
        decoded_word += decoded_letter
    
    decoded_message += (decoded_word) + " "

print({"Decoded Message:", decoded_message})

{'20 8 5  16 1 20 20 5 18 14  9 19  20 15  3 12 5 1 14  21 16  20 8 5  6 9 12 5  2 25  18 5 13 15 22 9 14 7  1 6 20 5 18  5 1 3 8  22 15 23 5 12  20 8 5  19 25 13 2 15 12 19  16 5 18 3 5 14 20  4 15 12 12 1 18  3 1 18 18 15 20  1 20', 'Encoded Message:'}
{'the pattern is to clean up the file by removing after each vowel the symbols percent dollar carrot at ', 'Decoded Message:'}


**TASK 4.3 (10 Points):**

**Restore** the interrogation statements back to their original state after **following the decoded message from before**.
- After restoration, make sure your modifications are updated in the **interrogation_df** itself.
- **Display** the **interrogation_df** when you are done.

In [582]:
import re

def clean_statement(statement):
    pattern = f"([aeiouAEIOU])[%$^@]"
    return re.sub(pattern, r"\1", statement)

interrogation_df['Statement'] = interrogation_df['Statement'].apply(clean_statement)

pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows', None)
interrogation_df

Name  \
0         Kristopher Stoichkov   
1                Naman Nagelia   
2         Pranavh Vallabhaneni   
3                    Stacy Sun   
4               Rina Kobayashi   
5                Asmita Brahme   
6                  Vismay Igur   
7                      Hted Oo   
8                Ankit Nakhawa   
9                       Ivy Lu   
10                  Danny Pham   
11                   Eric Chen   
12                Viktor Rakov   
13                 Min-seo Kim   
14              Giulia Bianchi   
15               Franco Romero   
16                Raymond Chen   
17                     Qiyu An   
18                 Jonah Chang   
19             Aashrit Sabbani   
20             Seungcheol Choi   
21                 Aidan Myers   
22              Amanvir Parhar   
23               Levi Ackerman   
24                  Jisoo Hong   
25                   Jake Farr   
26                    Andy Sun   
27                  David Guan   
28            Francesco Romano   
29                Pulkit Kumar   
30                  Neel Kumar   
31             Archer Sariscak   
32               Raha Moshasha   
33             Brendyn Royster   
34                 Suhan Neema   
35               James Johnson   
36                    Emily Li   
37            Thibaut Courtois   
38              Christina Jung   
39               Spencer Zheng   
40         Noah Der Garabedian   
41          Shashank Thirumale   
42                  Maheen Gul   
43                 George Witt   
44          Hansol Vernon Chwe   
45                   Jacob Liu   
46                  Austin Luu   
47                 Danhong Wei   
48                  Tony Zhang   
49               Thomas Nguyen   
50                 Tanvi Panse   
51               Saxon Wolcott   
52              Timothy Redman   
53              Soonyoung Kwon   
54      Adedapo Oyewole-Magnus   
55                Madison Beer   
56                 Emily Cheng   
57                 Rehan Kabir   
58              Kevin Mitchell   
59               Vrishin Patel   
60                  Emily Chen   
61             Nathaniel Green   
62                    Kai Chen   
63             Anushk Pokharna   
64                Winona Ryder   
65                 Nandita Das   
66                    Xiao Mei   
67                 Andrew Ross   
68               Pranav Sridar   
69             Thomas Anderson   
70            Valentina Torres   
71              Adrian Alphons   
72                Hye-jin Kang   
73              Shashank Arava   
74              Trinity Martin   
75            Shreyaj Bhandari   
76                Aidan Caruso   
77                Kwame Mensah   
78            Ananya Malipeddi   
79                  Jihoon Lee   
80                   Jason Liu   
81              Shayan Sobhani   
82               Young-min Seo   
83           Benjamin Talesnik   
84         Dimitrios Angelakos   
85               Jonathan Roth   
86               Jeonghan Yoon   
87                 Joseph Tang   
88                 Susan White   
89                 Valerie Lin   
90               Jamal Bankole   
91                  Eshan Khan   
92                  Minghao Xu   
93               Aneesa Ayornu   
94            Manasvi Sriranga   
95                   Ethan Qiu   
96                 Arjun Mudda   
97            Giovanni Moretti   
98               Jigar Bhavsar   
99               Anushree Daga   
100            Avanti Athavale   
101               Ella Banting   
102              Rakhee Gulzar   
103                Aditya Koul   
104                Seokmin Lee   
105              Lucia Morales   
106                  Chen Jing   
107                 Abdul Khan   
108             Abhyuday Goyal   
109                Vinay Gundu   
110              Erica Honadel   
111              Krish Thakker   
112                 Xinyu Yang   
113                  Somy Park   
114                  Hellen Ou   
115            Sakura Nakamura   
116        Achyutha Duddebanda   
117           


**TASK 4.4 (8 Points):** Create a column called **Ticket_Color** for the **"show_df"** that displays all the colors of the ticket

Remember that there was a show ticket left at the crime scene, take a second to look through **"show_df"**.
- Try and find a way to identify how you can find the ticket color from the ticket number.
- **Display** the **"show_df"** when you are done

**HINT:** Examine the **"interrogation_df"**, someone might have an idea

In [589]:
def get_ticket_color(ticket_number):
    digit_sum = sum(int(digit) for digit in str(ticket_number))
    rightmost_digit = digit_sum % 10
    
    if 0 <= rightmost_digit <= 2:
        return "Blue"
    elif 3 <= rightmost_digit <= 5:
        return "Red"
    elif 6 <= rightmost_digit <= 7:
        return "Green"
    elif 8 <= rightmost_digit <= 9:
        return "Yellow"

show_df['Ticket_Color'] = show_df['Ticket_Number'].apply(get_ticket_color)

show_df

ID  Ticket_Number Ticket_Color
0    P132       50269975          Red
1    P056       28120732          Red
2    P250       88051289         Blue
3    P197       79545169        Green
4    P012       83521845        Green
5    P033       22058855          Red
6    P021        9473450         Blue
7    P001       10116272         Blue
8    P249       23524843         Blue
9    P030       71886930         Blue
10   P071       92962076         Blue
11   P025       31002643       Yellow
12   P111       87338906          Red
13   P225       69020101       Yellow
14   P267       78443546         Blue
15   P259       28602637          Red
16   P265       72721943          Red
17   P105       94251199         Blue
18   P092       44345212          Red
19   P088       30175075       Yellow
20   P095         386818          Red
21   P239       92729414       Yellow
22   P242       42049173         Blue
23   P176       99491312       Yellow
24   P133        1659823          Red
25   P189       47856259        Green
26   P200       10043610          Red
27   P159       54158103        Green
28   P178        8631845          Red
29   P300       56763889         Blue
30   P138       43995703         Blue
31   P223       23480518         Blue
32   P023       45894659         Blue
33   P183       44607251       Yellow
34   P029       63045332        Green
35   P136        3439786         Blue
36   P114       84479148          Red
37   P052       44564514          Red
38   P160       47060216        Green
39   P166       66168819          Red
40   P031       94482194         Blue
41   P224       57655493          Red
42   P011       79781555        Green
43   P226        1263570          Red
44   P192       10310889         Blue
45   P038       34783723        Green
46   P284       42427144       Yellow
47   P282       96564203          Red
48   P098       48815710          Red
49   P298       37592884        Green
50   P051       56077968       Yellow
51   P256        7151664         Blue
52   P185       62339971         Blue
53   P203       43450497        Green
54   P067       74753690         Blue
55   P018       72179022         Blue
56   P150        5235256       Yellow
57   P247       31737732          Red
58   P080       41444365         Blue
59   P173       10481763         Blue
60   P143       59598056        Green
61   P172       69729467         Blue
62   P240       34267789        Green
63   P165       63669444         Blue
64   P102       59682930         Blue
65   P100       73118170       Yellow
66   P258        2353698        Green
67   P287       63892793        Green
68   P152       64066630         Blue
69   P162       61489246         Blue
70   P227       51075416       Yellow
71   P118       13605535       Yellow
72   P167        1823016         Blue
73   P024       72582055          Red
74   P062       27456284       Yellow
75   P175       86519622       Yellow
76   P026       89744345          Red
77   P131       61367768          Red
78   P221        1356101        Green
79   P123       88899885          Red
80   P010       45362017       Yellow
81   P019       45997909         Blue
82   P048       90981851         Blue
83   P116       92793637        Green
84   P151       92379415         Blue
85   P017       73305766        Green
86   P135       54314497        Green
87   P069        4820863         Blue
88   P297       55710602        Green
89   P110       66733075        Green
90   P217       68753837        Green
91   P211       71389135        Green
92   P235       87775511         Blue
93   P286       45056611       Yellow
94   P230       64380140        Green
95   P139       63745130       Yellow
96   P045       73546805       Yellow
97   P072       91344312        Green
98   P081       30425029          Red
99   P059       41254609         Blue
100  P248       74928644          Red
101  P099       45718117          Red
102  P041       58514994          Red
103  P039       56749629       Yellow
104  P009       8

## Q5) Finding the culprit (12 PTS Total)

A helpful function might be the "iloc" function. Have a look at: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html

**TASK 5.1(10 Points):** Using all the **data compiled from before** find a person that matches all the features from the profile.

- Do not hardcode or you will lose credit after all your hardwork 🥸

- Put the name of the culprit in a variable called **"culprit"**
- Print out the name of the culprit (There is only one).

In [596]:
potential_eye_colors = ['green', 'blue', 'orange', 'violet']
potential_hair_colors = ['green', 'blue', 'orange', 'violet']
potential_skin_colors = ['green', 'orange', 'red']
smallest_height = 150
largest_height = 170
smallest_age = 21
largest_age = 35
crime_location = "second floor"
ticket_color = "Red"

profiles_df = profiles_df.merge(show_df, on='ID', how='left')

culprit_profile = profiles_df[
    (profiles_df['Eye_Color'].isin(potential_eye_colors)) &
    (profiles_df['Hair_Color'].isin(potential_hair_colors)) &
    (profiles_df['Skin_Color'].isin(potential_skin_colors)) &
    (profiles_df['Height'] >= smallest_height) &
    (profiles_df['Height'] <= largest_height) &
    (profiles_df['Age'] >= smallest_age) &
    (profiles_df['Age'] <= largest_age) &
    (profiles_df['Location_during_crime'] == crime_location) &
    (profiles_df['Ticket_Color'] == ticket_color)
]

culprit = culprit_profile['Name'].values[0]

print(f"The culprit is: {culprit}")

The culprit is: LeBron James


**TASK 5.2 (2 Points):** Check if your culprit is correct!

**Interrogate** your culprit to see if you got the correct person.

In [600]:
statement = interrogation_df.loc[interrogation_df['Name'] == "LeBron James", 'Statement'].values[0]

print(f"LeBron James' Statement: {statement}")

LeBron James' Statement: "Forgive me sunshine. The markers were just too dang cool. Which court do I go, basketball or judicial?"


# GREAT WORK DETECTIVE!! 🌻